pkg/controller/customresource/customresource.go

```diff
@@ -139,3 +140,34 @@ func SetAnnotation(resource api.AtlasCustomResource, key, value string) {
 	annot[key] = value
 	resource.SetAnnotations(annot)
 }
+
+func ComputeSecretWithFallback[R api.ResourceWithCredentials](fallback bool, project *akov2.AtlasProject, resource R) (*client.ObjectKey, error) {
+	secret, err := resolveConnectionSecret(project, resource)
+	if fallback {
+		return secret, nil
+	}
+	return secret, err
```

Comment:
Let's not return objects and errors. The idiom is do mutually exclusively return either the result or the error, not both.